### ISSUE TITLE
cogvlm-chat模型的微调怎么实现呢？

### ISSUE THREAD

**Issue by xuxuxuchen** [THUDM/CogVLM#63]
我看到了finetune_224(490)_lora.sh的微调代码，但是没有看到cogvlm-chat微调代码，请问cogvlm-chat的微调代码应该如何实现呢？另外cogvlm-chat微调数据的方式是QA对还是什么样的呢？微调的lora中是仅微调语言模型的还是也包括视觉专家模型的呢？

**Comment by 1049451037** [comment]
Change these two lines:

https://github.com/THUDM/CogVLM/blob/e39f18652704737f1193b4939e4d40ec3a0872b1/scripts/finetune_224_lora.sh#L11-L12

into:

```
MODEL_TYPE="cogvlm-chat"
VERSION="chat"
```

Then you can fine-tune chat.

**Comment by xuxuxuchen** [comment]
明白了，那输入的数据要调整到什么格式呢？我看给的base的finetune的输入数据是图像名称为label的形式，修改为chat后输入数据也是这样的吗？

**Comment by 1049451037** [comment]
数据格式你可以自己对接，修改dataset.py对应的代码：

https://github.com/THUDM/CogVLM/blob/e39f18652704737f1193b4939e4d40ec3a0872b1/utils/dataset.py#L23-L51

**Comment by xuxuxuchen** [comment]
好的，明白了，微调chat的话就是修改代码中对应的img、prompt（问题）、answer（标准回答）是吗？

**Comment by xuxuxuchen** [comment]
比如现在的line53:text_dict = self.process_text(label, "CAPTCHA:")，我需要将"CAPTCHA:"换为对应的问题就可以了对吗？

**Comment by 1049451037** [comment]
是的

**Comment by pf67** [comment]
要改几个地方，第一个是sh文件，里面设定了你的数据集路径，模型版本，模型路径，还有参数大小等（比如哪个最大文本长度，默认的太小，必然会导致微调失败，你得改大一点），然后就是dataset里面，默认的label是读取文件名，只适合于验证码的那个数据集，你得根据自己需要修改，然后说prompt，和label一起传给text处理模块的，这几个改完基本就可以跑了，然后就是.json，那个是deepspeed的配置文件，如果你的卡不够多的话建议也要修改下batchsize这些参数。

**Comment by pf67** [comment]
其实490和224的配置基本没什么区别，就是mpsize改了下，但是注意224里读取的ds的配置还把batchsize改成了32，爆显存注意。
我微调是用的490sh，把mp改成4，batchsize改成3，4张4090基本跑满。

**Comment by microhu** [comment]
> 比如现在的line53:text_dict = self.process_text(label, "CAPTCHA:")，我需要将"CAPTCHA:"换为对应的问题就可以了对吗？

如果微调的训练数据包含有多轮对话，这个需要怎么修改呢？目前是支持的吗？ 另外，如果还有多张图片呢 @1049451037

**Comment by 1049451037** [comment]
多轮对话也要整合成一个prompt和label，至于怎么整合可以自己写一个拼接操作。

多张图片目前还没有实现，但是模型本身是支持的，我暂时也没有一个特别完美优雅的解决方案。可以先自己随便写一个。

**Comment by MengXinChengXuYuan** [comment]
@1049451037 感谢开源先！
请教个问题，我想尝试一下在医疗数据上进行finetue，实现输入病人多张CT图，身份信息、年龄性别等相关信息，每轮输入一张CT图让网络进行描述，最终根据所有输入最终得出病人患病与否的结论
想请教下网络能记住以往多轮输入的特征（尤其是图像）吗？还是每一轮都是独立的？
感谢！！！！

**Comment by 1049451037** [comment]
参考这段话：

> 多轮对话也要整合成一个prompt和label，至于怎么整合可以自己写一个拼接操作。
> 
> 多张图片目前还没有实现，但是模型本身是支持的，我暂时也没有一个特别完美优雅的解决方案。可以先自己随便写一个。

**Comment by Rickylht** [comment]
> @1049451037 感谢开源先！ 请教个问题，我想尝试一下在医疗数据上进行finetue，实现输入病人多张CT图，身份信息、年龄性别等相关信息，每轮输入一张CT图让网络进行描述，最终根据所有输入最终得出病人患病与否的结论 想请教下网络能记住以往多轮输入的特征（尤其是图像）吗？还是每一轮都是独立的？ 感谢！！！！

这也是我正在研究的场景。我觉得制作一个质量非常高的数据集真的会更加重要。多轮多图实在是太复杂了。多图单轮会好一点。多图多轮有deepspeedvisual那一篇，经过复现效果也不好而且代码也有bug，目前没什么好的解决方案。不如用llava多图单轮作为baseline，我感觉会好调很多。

**Comment by jyC23333** [comment]
@1049451037 @pf67 请教下finetune的时候验证集loss一直如下，这是正常的吗？

![image](https://github.com/THUDM/CogVLM/assets/110331827/bf5e5235-f727-493d-9710-d04ce9d0098f)

**Comment by 1049451037** [comment]
如果只训练yes，模型还能坚持输出No.（还是大写开头）只能说明可训练参数太少了。

这里可以配置可训练参数有哪些：

https://github.com/THUDM/CogVLM/blob/3c254565408e180207de8a0b69f20473ecd5c157/finetune_demo.py#L13-L36

**Comment by pf67** [comment]
> @1049451037 @pf67 请教下finetune的时候验证集loss一直如下，这是正常的吗？
> 
> ![image](https://user-images.githubusercontent.com/110331827/283741306-bf5e5235-f727-493d-9710-d04ce9d0098f.png)

不正常，首先你应该先关注一下训练集在训练过程中的loss是个什么水平。从验证集结果看，你的验证集上label为小写的yes，而模型输出结果为首字母大写的Yes.或No.，出现这种差异性，应该先检查一下样本的标签是不是一致。

**Comment by hzauliutingting** [comment]
![QQ截图20231123132123](https://github.com/THUDM/CogVLM/assets/95989981/d76658f5-b440-457f-9cfc-7d36a38b6432)
我在微调的时候把MODEL_TYPE变量改为自己下载的模型路径，然后运行的时候出现xia'mia下面的错误，请问这样该怎么解决呢
![QQ截图20231123132214](https://github.com/THUDM/CogVLM/assets/95989981/a193738c-055d-4af9-9afa-f49277859b8e)

**Comment by hzauliutingting** [comment]
> ![QQ截图20231123132123](https://user-images.githubusercontent.com/95989981/285106311-d76658f5-b440-457f-9cfc-7d36a38b6432.png) 我在微调的时候把MODEL_TYPE变量改为自己下载的模型路径，然后运行的时候出现xia'mia下面的错误，请问这样该怎么解决呢 ![QQ截图20231123132214](https://user-images.githubusercontent.com/95989981/285106401-a193738c-055d-4af9-9afa-f49277859b8e.png)

另外我还把修改了变量：
--local_tokenizer vicuna-7b-v1.5 \  
vicuna-7b-v1.5是我下载到当前运行路径下文件名称

**Comment by 1049451037** [comment]
截图不全，我也看不出来

**Comment by hzauliutingting** [comment]
> 截图不全，我也看不出来。但总之就是改错了，导致bash脚本损坏了。
> 
> 检查一下是不是哪一行最后的 \ 丢了

非常感激您的回复！
我按照原来的代码运行也会出现这样的错误，在那之前，我有以下的一些错误，我想是因为linux系统和window系统编写的格式问题，所以我把空行去掉了。
![QQ截图20231123140904](https://github.com/THUDM/CogVLM/assets/95989981/7fe87c14-3e91-4db3-b801-9c2c043558f4)
但是无论我怎么修改，还是下面的错误。(https://github.com/THUDM/CogVLM/assets/95989981/58d17380-a87b-4d57-b9f9-fa8dead2634d)

**Comment by 1049451037** [comment]
用一个功能多一点的文本编辑器，比如vscode，把\r都替换成\r\n

**Comment by hzauliutingting** [comment]
> 用一个功能多一点的文本编辑器，比如vscode，把\r都替换成\r\n

您好，我用的就是vscode编辑器，我是直接把空行去掉，不再报错$‘r’:未找到命令，但是还是报错error: unrecognized arguments: 
![QQ截图20231123143627](https://github.com/THUDM/CogVLM/assets/95989981/2982653a-6d85-41c5-8cf7-54abf9e88c22)

**Comment by 1049451037** [comment]
所以说不要去掉空行，而是把\r都替换成\r\n。因为有的\r你看不到，但是还是存在。

**Comment by hzauliutingting** [comment]
> 所以说不要去掉空行，而是把\r都替换成\r\n。因为有的\r你看不到，但是还是存在，就导致了unrecognized问题。

好的，我试试，非常感谢~

**Comment by 1049451037** [comment]
是的，如果要评测其他任务，相应的指标需要您自己实现。

**Comment by wulouzhu** [comment]
> 是的，如果要评测其他任务，相应的指标需要您自己实现。
我理解项目中包含了图片描述的指标，应该有图片描述的评估代码吧？你能帮我指出图片描述评估代码的位置吗？谢谢

**Comment by pf67** [comment]
> @jyC23333 你好，我也遇到了这个问题，我的标签是一段描述，但微调过程中验证集的loss一直是0，训练集的loss是在下降的，请问你解决了吗？ @pf67 @1049451037 另外，我看evaluate_demo.py的评估指标是直接判断label和pred是否相等，如果label是图片描述的话，这种评估方法是不是有问题？谢谢 ![image](https://private-user-images.githubusercontent.com/31339491/288956025-4493169e-4542-47c2-84dc-b46c558b6cc1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDIwMDc5MjMsIm5iZiI6MTcwMjAwNzYyMywicGF0aCI6Ii8zMTMzOTQ5MS8yODg5NTYwMjUtNDQ5MzE2OWUtNDU0Mi00N2MyLTg0ZGMtYjQ2YzU1OGI2Y2MxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjA4VDAzNTM0M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA5N2RhOGYzMjg3Zjg3MmVjNTMxY2YxZDJhNWZlZDc2YTRlODJiZWU1NmFjNzkzNzQ2MWJmMjI2ZWM0NDY1MTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.GdX0UT-EVlv2KQvfUmviI23E6hm7xgMfeJyo-BqcwJ8)
因为官方给的样例是一个相对简单的OCR字符识别场景，输出是固定的，所以他们用的这个评估方式。
但是实际对生成式LLM的应用，这么简单的评估方式显然是没有通用性的，尤其是你说的对于多模特图片识别上，据我所知业界也没有一个通用的方案，建议根据自己的实际需求去做评估吧。

**Comment by wulouzhu** [comment]
> > @jyC23333 你好，我也遇到了这个问题，我的标签是一段描述，但微调过程中验证集的loss一直是0，训练集的loss是在下降的，请问你解决了吗？ @pf67 @1049451037 另外，我看evaluate_demo.py的评估指标是直接判断label和pred是否相等，如果label是图片描述的话，这种评估方法是不是有问题？谢谢 ![image](https://private-user-images.githubusercontent.com/31339491/288956025-4493169e-4542-47c2-84dc-b46c558b6cc1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDIwMDc5MjMsIm5iZiI6MTcwMjAwNzYyMywicGF0aCI6Ii8zMTMzOTQ5MS8yODg5NTYwMjUtNDQ5MzE2OWUtNDU0Mi00N2MyLTg0ZGMtYjQ2YzU1OGI2Y2MxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjA4VDAzNTM0M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA5N2RhOGYzMjg3Zjg3MmVjNTMxY2YxZDJhNWZlZDc2YTRlODJiZWU1NmFjNzkzNzQ2MWJmMjI2ZWM0NDY1MTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.GdX0UT-EVlv2KQvfUmviI23E6hm7xgMfeJyo-BqcwJ8)
> > 因为官方给的样例是一个相对简单的OCR字符识别场景，输出是固定的，所以他们用的这个评估方式。
> > 但是实际对生成式LLM的应用，这么简单的评估方式显然是没有通用性的，尤其是你说的对于多模态图片识别上，据我所知业界也没有一个通用的方案，建议根据自己的实际需求去做评估吧。

谢谢答复！另外，再请教一个问题，我想将web服务的host改成0.0.0.0, 在终端将MASTER_ADDR变量设为0.0.0.0好像不起作用，请问该如何改，感谢~

**Comment by wulouzhu** [comment]
> > > @jyC23333 你好，我也遇到了这个问题，我的标签是一段描述，但微调过程中验证集的loss一直是0，训练集的loss是在下降的，请问你解决了吗？ @pf67 @1049451037 另外，我看evaluate_demo.py的评估指标是直接判断label和pred是否相等，如果label是图片描述的话，这种评估方法是不是有问题？谢谢 ![image](https://private-user-images.githubusercontent.com/31339491/288956025-4493169e-4542-47c2-84dc-b46c558b6cc1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDIwMDc5MjMsIm5iZiI6MTcwMjAwNzYyMywicGF0aCI6Ii8zMTMzOTQ5MS8yODg5NTYwMjUtNDQ5MzE2OWUtNDU0Mi00N2MyLTg0ZGMtYjQ2YzU1OGI2Y2MxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjA4VDAzNTM0M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA5N2RhOGYzMjg3Zjg3MmVjNTMxY2YxZDJhNWZlZDc2YTRlODJiZWU1NmFjNzkzNzQ2MWJmMjI2ZWM0NDY1MTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.GdX0UT-EVlv2KQvfUmviI23E6hm7xgMfeJyo-BqcwJ8)
> > > 因为官方给的样例是一个相对简单的OCR字符识别场景，输出是固定的，所以他们用的这个评估方式。
> > > 但是实际对生成式LLM的应用，这么简单的评估方式显然是没有通用性的，尤其是你说的对于多模态图片识别上，据我所知业界也没有一个通用的方案，建议根据自己的实际需求去做评估吧。
> 
> 谢谢答复！另外，再请教一个问题，我想将web服务的host改成0.0.0.0, 在终端将MASTER_ADDR变量设为0.0.0.0好像不起作用，请问该如何改，感谢~

已经找到了，谢谢
